### PR TITLE
Move and rewrite version identifier

### PR DESCRIPTION
**Pull request by mmitche**
This change is two parts:
- Move the version identifier from arcade-services - This is being done because aka.ms link generation during publishing needs to remove version numbers, and that would mean that both the maestro push task in arcade-services and the publishing tasks in arcade would have a dependency on the version identifier package, built out of arcade-services. For layering purposes, it's better to move this into arcade (where other asset manifest code lives anyway).
- To properly generate stable aka.ms links during the publishing process, we need to be able to remove all version numbers from blob assets. The current version identifier code incorrectly identifies certain cases (e.g. nuget packages that end in a version number like "2.1", and also tends to preference the version number in a path segment over the version number in a file name.  I rewrote this code to be correct in those cases and added an API that will remove all version numbers.

For testing purposes, I've replaced the existing test list with all the unique asset names that exist within the BAR database, and added an extra column for what the asset would look like without any version numbers.

### CODE REVIEW

**Comment by mmitche** [comment]
This work is required prior to https://github.com/dotnet/arcade/issues/3963 continuing

**Comment by dagood** [comment]
Why is this necessary to generate latest links? Is this a symptom of some data being missing in BAR--shouldn't the repo be responsible for knowing what versions it produced and passing that along?

Extracting version from filename is complex and creates potentially unknown assumptions, and I recommend never doing it. It's somewhat more understandable further downstream where there isn't influence over the build, but from what I know now it seems reasonable to require repos to pass along the version to activate Arcade's latest link generation infra.

Since a link will be generated for every single artifact automatically, a break here means that the infra will start generating a new aka.ms link for every build, until someone notices. (And AFAIK there is no reason anyone would notice, because there is no reason to use the link for many of these artifacts.) Is it particularly bad to load the link service like that, with creates rather than updates?

At least with this impl, unlike others, there are tests for what we have now--so only new stuff may break. If it's necessary to get latest links generated by Arcade in the short term without other build changes (why?) then I'm ok merging it.

**Comment by mmitche** [comment]
> Why is this necessary to generate latest links? Is this a symptom of some data being missing in BAR--shouldn't the repo be responsible for knowing what versions it produced and passing that along?
> 
> Extracting version from filename is complex and creates potentially unknown assumptions, and I recommend never doing it. It's somewhat more understandable further downstream where there isn't influence over the build, but from what I know now it seems reasonable to require repos to pass along the version to activate Arcade's latest link generation infra.
> 
> Since a link will be generated for every single artifact automatically, a break here means that the infra will start generating a new aka.ms link for every build, until someone notices. (And AFAIK there is no reason anyone would notice, because there is no reason to use the link for many of these artifacts.) Is it particularly bad to load the link service like that, with creates rather than updates?
> 
> At least with this impl, unlike others, there are tests for what we have now--so only new stuff may break. If it's necessary to get latest links generated by Arcade in the short term without other build changes (why?) then I'm ok merging it.

This is what we do currently, just that the implementation is wrong and this is correcting it.  Note that for nuget packages we actually use the nuget metadata when we detect version numbers.  I do not believe that the version numbers for blobs are in use at all today. The repositories probably can identify version numbers properly, but quite a bit of work may be required in various places to feed the data through.  Most of the time by the time publishing happens, we simply have a list of files.

What we could do would be require that repos provide extra metadata for each blob they want to generate a link for, indicating what the stable link should look like.  For repos like core-setup, this would **still** mean either recovering version data during signing + publishing, or feeding it across the build job to the signing job. You'll see that even today, there is somewhat interesting logic required to double publish files with different version numbers. 

So basically it might come down to a tradeoff:
- Generate an akams link automatically using our "guesses". Yes, maybe at some point something changes and we end up generating links that change over time. Generally, these links should be fine as long as we don;'t move from semver2 **or**
- Implement logic in each repo that wants akams links to strip away/identify version numbers (remember that multiple different version numbers may be present in a single blob path) and pass that data along.  This is likely significant amounts of work in some repos (core-setup comes to mind), and potentially just as error prone.
-

**Comment by dagood** [comment]
> but quite a bit of work may be required in various places to feed the data through.

> For repos like core-setup, this would still mean either recovering version data during signing + publishing, or feeding it across the build job to the signing job. You'll see that even today, there is somewhat interesting logic required to double publish files with different version numbers.

Version calculation based on build number is more reliable/comprehensible now than it was on BuildTools, so I'd say the current impl of this stuff in Core-Setup is a relic and worth replacing. However, with other repos being in the same situation (maybe worse?) I think this is a reasonable argument for parsing filenames to let everyone use the new infra without rework.

I'll review the PR in a bit.

File: src/Microsoft.DotNet.VersionTools/tests/Microsoft.DotNet.VersionTools.Tests.csproj [hunk 1]
```diff
@@ -12,4 +12,10 @@
     <ProjectReference Include="..\lib\Microsoft.DotNet.VersionTools.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <None Update="BuildManifest\VersionIdentifierTestsAssets.csv">
```
Comment (dagood) [inline src/Microsoft.DotNet.VersionTools/tests/Microsoft.DotNet.VersionTools.Tests.csproj]: Maybe grab historical test data from dotnet/versions too? `Last_Build_Packages.txt`s have `{id} {version}` so I think it would be relatively easy to put in this format.

File: src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            {
+                { "-.", "." },
+                { "..", "." },
+                { "--", "-" },
+                { "//", "/" },
+                { "_.", "." }
+            };
+
+        /// <summary>
+        ///     Given an asset name, remove all .NET Core version numers (as defined by the version identifier above)
```
Comment (dagood) [inline src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs]:
```suggestion
        ///     Given an asset name, remove all .NET Core version numbers (as defined by GetVersionForSingleSegment)
```

File: src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+            if (string.IsNullOrEmpty(majorMinorPatch))
+            {
+                return null;
+            }
+
+            return $"{majorMinorPatch}{versionSuffix.ToString()}";
+        }
+
+        private static readonly Dictionary<string, string> sequencesToReplace =
+            new Dictionary<string, string>
```
Comment (dagood) [inline src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs]: I would use something ordered rather than a dictionary to make sure the results are consistent.

File: src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+                // Don't use else, so that we don't miss segments
+                // in case we are just deciding that we've finished major minor patch.
+                if (majorMinorPatch != null)
+                {
+                    // Now look at the next segment. If it looks like it could be part of a version. If it can't, then
+                    // we're done.
+                    // Cases where we should break out and be done:
+                    // - We have an empty pre-release label and the delimiter is not '-'.
+                    // - We have an empty pre-release label and the next segment does not start with a know tag.
+                    // - We have a non-empty pre-release label and the current segment is not a number.
```
Comment (dagood) [inline src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs]: Should `final` be mentioned in this list? It seems like `we should break out and be done [if] We have a non-empty pre-release label and the current segment is not a number.` is wrong, since `final` isn't a number.

File: src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+                
+                // Don't use else, so that we don't miss segments
+                // in case we are just deciding that we've finished major minor patch.
+                if (majorMinorPatch != null)
+                {
+                    // Now look at the next segment. If it looks like it could be part of a version. If it can't, then
+                    // we're done.
+                    // Cases where we should break out and be done:
+                    // - We have an empty pre-release label and the delimiter is not '-'.
+                    // - We have an empty pre-release label and the next segment does not start with a know tag.
```
Comment (dagood) [inline src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs]:
```suggestion
                    // - We have an empty pre-release label and the next segment does not start with a known tag.
```

File: src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+                        int major = majorMinorPatchStack.Pop();
+                        majorMinorPatch = $"{major}.{minor}.{patch}";
+                    }
+                }
+                
+                // Don't use else, so that we don't miss segments
+                // in case we are just deciding that we've finished major minor patch.
+                if (majorMinorPatch != null)
+                {
+                    // Now look at the next segment. If it looks like it could be part of a version. If it can't, then
```
Comment (dagood) [inline src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs]: Not sure:

```suggestion
                    // Now look at the next segment. If it looks like it could be part of a version, continue. If it can't, then
```

File: src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+
+            if (string.IsNullOrEmpty(majorMinorPatch))
+            {
+                return null;
+            }
+
+            return $"{majorMinorPatch}{versionSuffix.ToString()}";
+        }
+
+        private static readonly Dictionary<string, string> sequencesToReplace =
```
Comment (dagood) [inline src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs]: Style issue: should be near top of file and use `_` prefix.

File: src/Microsoft.DotNet.VersionTools/tests/BuildManifest/VersionIdentiferTests.cs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            foreach (VersionIdentifierTestAsset testAsset in testAssets)
+            {
+                // First check whether the original version number can be identified
+                string expectedVersion = testAsset.ExpectedVersion;
+                string actualVersion = VersionIdentifier.GetVersion(testAsset.Name);
+                Assert.True(expectedVersion == actualVersion, $"Line {testAsset.Line} has incorrect computed version {actualVersion}");
+                // Then check that all versions can be removed from the path of any blob asset
+                string expectedNameWithoutVersions = testAsset.NameWithoutVersions;
+                string actualNameWithoutVersions = VersionIdentifier.RemoveVersions(testAsset.Name);
+                Assert.True(expectedNameWithoutVersions == actualNameWithoutVersions, $"Line {testAsset.Line} has incorrect asset name with versions removed {actualNameWithoutVersions}");
```
Comment (JohnTortugo) [inline src/Microsoft.DotNet.VersionTools/tests/BuildManifest/VersionIdentiferTests.cs]: NIT: maybe "has incorrect asset `NameWithoutVersions`"?

File: src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+            };
+
+        private const string _finalSuffix = "final";
+
+        private static readonly char[] _delimiters = new char[] { '.', '-', '_' };
+
+        /// <summary>
+        /// Identify the version of an asset.
+        /// 
+        /// Assets can come in two forms:
```
Comment (JohnTortugo) [inline src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs]: NIT: "Assets' name"?

File: src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+        /// - Packages that do not include any path elements.
+        /// 
+        /// There may be multiple different version numbers in a blob path.
+        /// This method starts at the last segment of the path and works backward to find a version number.
+        /// </summary>
+        /// <param name="assetName">Asset name</param>
+        /// <returns>Version, or null if none is found.</returns>
+        public static string GetVersion(string assetName)
+        {
+            string[] pathSegments = assetName.Split(new char[] { '/' }, StringSplitOptions.RemoveEmptyEntries);
```
Comment (JohnTortugo) [inline src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs]: Don't we need to care about "\" too?

File: src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+            string potentialVersion = null;
+            for (int i = pathSegments.Length - 1; i >= 0; i--)
+            {
+                potentialVersion = GetVersionForSingleSegment(pathSegments[i]);
+                if (potentialVersion != null)
+                {
+                    break;
```
[the remark this answers is not in the document]
Comment (JohnTortugo) [reply src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs]: Agree with changing the `break` to `return` here.

File: src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+        /// </summary>
+        /// <param name="assetSegment">Asset segment</param>
+        /// <returns>Version number, or null if none was found</returns>
+        /// <remarks>
+        /// Identifying versions is not particularly easy. To constrain the problem, we apply the following assumptions
+        /// which are generally valid for .NET Core.
+        /// - We always have major.minor.patch, and it always begins the version string.
+        /// - The only pre-release or build metadata labels we use begin with the _knownTags shown above.
+        /// - We use additional numbers in our version numbers after the initial major.minor.patch, but any non-numeric element will end the version string
+        /// - The delimiters we use in versions and file names are ., -, and _.
```
Comment (JohnTortugo) [inline src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs]: NIT: refer to `_delimiters` ?

File: src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+        /// </remarks>
+        private static string GetVersionForSingleSegment(string assetSegment)
+        {
+
+            // Find the start of the version number by finding the major.minor.patch.
+            // Scan the string forward looking for a digit preceded by one of the delimiters,
+            // then look for a minor.patch, completing the major.minor.patch.  Continue to do so until we get
+            // to something that is NOT major.minor.patch (this is necessary because we sometimes see things like:
+            // VS.Redist.Common.NetCore.Templates.x86.2.2.3.0.101-servicing-014320.nupkg
```
Comment (JohnTortugo) [inline src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs]: Didn't get to the logic below yet, but just curious how it will handle the ambiguous options. Is the version "2.2.3.0.101", "2.3.0.101", "3.0.101", etc. in the general case (i.e., no pre-release fields).

File: src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+                prevDelimiterCharacter = nextDelimiterCharacter;
+                int nextDelimiterIndex = assetSegment.IndexOfAny(_delimiters, currentIndex);
+                if (nextDelimiterIndex != -1)
+                {
+                    nextDelimiterCharacter = assetSegment[nextDelimiterIndex];
+                    nextSegment = assetSegment.Substring(currentIndex, nextDelimiterIndex - currentIndex);
+                }
+                else
+                {
+                    nextSegment = assetSegment.Substring(currentIndex, assetSegment.Length - currentIndex);
```
Comment (JohnTortugo) [inline src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs]: NIT: maybe just `assetSegment.Substring(currentIndex)` ?

**Comment by mmitche** [comment]
> Version calculation based on build number is more reliable/comprehensible now than it was on BuildTools, so I'd say the current impl of this stuff in Core-Setup is a relic and worth replacing. However, with other repos being in the same situation (maybe worse?) I think this is a reasonable argument for parsing filenames to let everyone use the new infra without rework.

@dagood I think the rest of the repos end up in the same situation (core-sdk has similar logic)

File: src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs (hunk 3 above)
Comment (mmitche) [reply src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs]: replaced with SortedDictionary

File: src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs (hunk 10 above)
Comment (mmitche) [reply src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs]: Nope.  They would not contain versions.

File: src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs (hunk 13 above)
Comment (mmitche) [reply src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs]: it will continue forward and pick 3.0.101 in all those cases, which is what we want for the packages that have this, like Templates.2.1.3.0.101, which is the 2.1 templates built out of the 3.0.1xx build.

**Comment by mmitche** [comment]
Feedback addressed.

**Comment by dagood** [comment]
I have 3 comments in a "3 hidden conversations" bubble with no response, did you hit those?

File: src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+                        pathSegments[i] = pathSegments[i].Replace(versionForSegment, "");
+                    }
+                }
+            }
+
+            string assetWithoutVersions = string.Join("/", pathSegments);
+
+            foreach (var sequence in sequencesToReplace)
+            {
+                assetWithoutVersions = assetWithoutVersions.Replace(sequence.Key, sequence.Value);
```
[the remark this answers is not in the document]
Comment (mmitche) [reply src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs]: > I have 3 comments in a "3 hidden conversations" bubble with no response, did you hit those?

Ahh, no...missed those.

File: src/Microsoft.DotNet.VersionTools/tests/Microsoft.DotNet.VersionTools.Tests.csproj (hunk 1 above)
Comment (mmitche) [reply src/Microsoft.DotNet.VersionTools/tests/Microsoft.DotNet.VersionTools.Tests.csproj]: Not a bad idea.  We shouldn't see really any difference here (the old semver1 versions of various repos early on in the BAR look like 2.x packages, but it's worth adding for completeness)

**Comment by mmitche** [comment]
Added some packages from the versions repo (no issues found), and altered the sequence replacement.
